rio/transmat/impl/git2/git_internals.go

```diff
@@ -0,0 +1,380 @@
+	"gopkg.in/src-d/go-git.v4/plumbing/filemode"
+	"gopkg.in/src-d/go-git.v4/plumbing/format/packfile"
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+	"gopkg.in/src-d/go-git.v4/plumbing/protocol/packp"
+	"gopkg.in/src-d/go-git.v4/plumbing/storer"
+	"gopkg.in/src-d/go-git.v4/plumbing/transport"
+	"gopkg.in/src-d/go-git.v4/plumbing/transport/client"
+	"gopkg.in/src-d/go-git.v4/storage/filesystem"
+	"gopkg.in/src-d/go-git.v4/storage/memory"
+	"gopkg.in/src-d/go-git.v4/utils/ioutil"
+)
+
+type filePlacer interface {
+	Place(*object.File) error
+}
+
+type gitFilePlacer struct {
+	billy.Filesystem
+}
+
+func (g *gitFilePlacer) Place(f *object.File) error {
+	return checkoutFile(f, g)
+}
+
+func newFilePlacer(fs billy.Filesystem) filePlacer {
+	return &gitFilePlacer{fs}
+}
+
+func placeTree(tree *object.Tree, fp filePlacer) {
+	fileIterator := tree.Files()
+	err := fileIterator.ForEach(fp.Place)
+	if err != nil {
+		panic(meep.Meep(
+			&rio.ErrInternal{Msg: "Unable to place git tree onto the file system"},
+			meep.Cause(err),
+		))
+	}
+}
+
+func gitLsRemote(url string) (memory.ReferenceStorage, error) {
+	endpoint, err := transport.NewEndpoint(url)
+	if err != nil {
+		return nil, err
+	}
+	gitClient, err := client.NewClient(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	gitSession, err := gitClient.NewUploadPackSession(endpoint, nil)
+	if err != nil {
+		return nil, err
+	}
+	advertisedRefs, err := gitSession.AdvertisedReferences()
+	if err != nil {
+		return nil, err
+	}
+	refs, err := advertisedRefs.AllReferences()
+	if err != nil {
+		return nil, err
+	}
+	err = gitSession.Close()
+	if err != nil {
+		return nil, err
+	}
+	return refs, nil
+}
+
+func gitCheckout(commit *object.Commit, fs billy.Filesystem) {
+	tree, err := commit.Tree()
+	if err != nil {
+		panic(meep.Meep(
+			&rio.ErrInternal{Msg: "Commit missing tree object"},
+			meep.Cause(err),
+		))
+	}
+	placeTree(tree, newFilePlacer(fs))
+}
+
+/*
+	Transform the rio commit ID to a git hash.
+	Performs some basic checks on inputs.
+*/
+func CommitId2Hash(hash rio.CommitID) plumbing.Hash {
+	mustBeFullHash(hash)
+	ref := plumbing.NewReferenceFromStrings("", string(hash))
+	return ref.Hash()
+}
+
+/*
+	Reads the gitmodules file and checks each entry against the commit tree entries.
+	Returns a map of submodule config objects to matching tree entries
+*/
+func listSubmodules(commit *object.Commit, fs billy.Filesystem, commitHash plumbing.Hash) map[*config.Submodule]*object.TreeEntry {
+	tree, err := commit.Tree()
+	if err != nil {
+		panic(meep.Meep(
+			&rio.ErrInternal{Msg: "Commit missing tree object"},
+			meep.Cause(err),
+		))
+	}
+
+	cfgModules, err := readGitmodulesFile(fs)
+	if err != nil {
+		panic(meep.Meep(
+			&rio.ErrInternal{Msg: "Unable to read gitmodules file"},
+			meep.Cause(err),
+		))
+	}
+
+	result := map[*config.Submodule]*object.TreeEntry{}
+	if cfgModules != nil {
+		for _, submodule := range cfgModules.Submodules {
+			if submodule == nil {
```

Comment:
This is guarding against... the `.gitmodules` file having incomplete entries, is that correct?

The `tar` transmat uses `def.ErrWareCorrupt` for things like this; that's probably applicable here too.  (Which I suddenly notice *isn't* in the [list of valid errors](https://github.com/polydawn/repeatr/blob/9be3ceed60b42096455e9ee92ccf41417aa5b105/rio/errors.go#L65-L74) to return from this area -- I'm going to fix that immediately in another branch because it's Clearly Wrong.)